### PR TITLE
multithreaded evalMinMax

### PR DESCRIPTION
**Pull request by ghurstunither**
Reimplements `evalMinMax` through `DynamicNodeManager` to enable threading.

The following example profiles `evalMinMax` over a sphere of radius 50, spacing 0.1, and halfwidth 3.0.

Before:
>     min: -0.299899, max: 0.299899
>     time: 1926.2ms

After:
>     min: -0.299899, max: 0.299899
>     time: 48.9077ms

Code used to generate the above outputs:

```
#include <openvdb/openvdb.h>
#include <openvdb/tools/LevelSetSphere.h>

#include <chrono>
#include <iostream>

int main()
{
    using openvdb::FloatGrid;
    using openvdb::tools::createLevelSetSphere;
    
    using std::chrono::high_resolution_clock;
    using std::chrono::duration;
    using std::chrono::milliseconds;
    
    openvdb::initialize();
    
    FloatGrid::Ptr grid = createLevelSetSphere<FloatGrid>(50.0, openvdb::Vec3f(0, 0, 0), 0.1, 3.0);

	float min, max;
	
    // profiled evaluation:
	auto t1 = high_resolution_clock::now();
    grid->evalMinMax(min, max);
    auto t2 = high_resolution_clock::now();
    
    duration<double, std::milli> ms_double = t2 - t1;

    std::cout << "min: " << min << ", max: " << max << std::endl;
    std::cout << "time: " << ms_double.count() << "ms" << std::endl;
    return 0;
}

```

### CODE REVIEW

File: openvdb/openvdb/tree/Tree.h [hunk 1]
```diff
@@ -361,7 +361,7 @@ class Tree: public TreeBase
     Index64 activeTileCount() const override { return tools::countActiveTiles(*this); }
 
     /// Return the minimum and maximum active values in this tree.
-    void evalMinMax(ValueType &min, ValueType &max) const;
+    void evalMinMax(ValueType &min, ValueType &max) const { tools::minMaxValues(*this, min, max); }
```
[the remark this answers is not in the document]
Comment (Idclip) [reply openvdb/openvdb/tree/Tree.h]: Thought about this a bit more, I don't think it's worth sticking with the single threaded option on `evalMinMax` if we're not also doing that for other methods such as `activeVoxelCount` as that just isn't consistent. So as long as we mark this method as deprecated then perhaps this change is fine. We can still discuss in TSC but let's get this in as is (+ the deprecation).

@ghurstunither I notice there's CI failures, are you able to take a look and fix those?

[the remark this answers is not in the document]
Comment (ghurstunither) [reply openvdb/openvdb/tree/Tree.h]: Thanks for pointing out the unit tests! I was not correctly supporting edge cases like empty grids. This should be fixed now.

[the remark this answers is not in the document]
Comment (ghurstunither) [reply openvdb/openvdb/tree/Tree.h]: Regardless of what's best here in this PR, _should_ `evalMinMax` and `activeVoxelCount`, etc introduce an optional threading argument? I'm not suggesting this would solve the problems referred to above, but is there harm adding this switch? It would be the easiest way for one to disable multithreading if desired.

[the remark this answers is not in the document]
Comment (danrbailey) [reply openvdb/openvdb/tree/Tree.h]: Given that we're looking to deprecate `Tree::evalMinMax()`, I would think that it's not necessary, but (and I think we'll talk about this in the next TSC), perhaps adding a comment to note that this method is now multi-threaded and to switch to use your new tools one if the single-threaded behavior is desirable?

**Comment by ghurstunither** [comment]
@Idclip @danrbailey 

I've gone ahead and deprecated `evalMinMax` and think it's ready for re-review. The deprecation process gave me a real-world reason to comb through and learn more about the entire code-base -- I'm grateful for the experience.

A few questions/comments:

* We agreed the most natural place for `tools::minMax` to live is in Statistics.h. However I think this causes a circular dependency since Statistics.h --> ValueTransformer.h --> Grid.h --> Tree.h and `tree->print()` now calls `tools::minMax` instead of `evalMinMax`. This would mean Tree.h would include Statistics.h, completing the cycle.
    * I've instead placed the implementation in Count.h, which other than similarity in implementation style isn't really the best place I think.
    * What's the best way to address this issue?
* I did not deprecate the python version of `evalMinMax` (in pyGrid.h) but changed the implementation to call `tools::minMax`. Is this ok?
* I removed the unit tests for `evalMinMax`, effectively migrating them and switching to `tools::minMax` in order to add tests for the new routine.
* All other calls to `evalMinMax` have been converted to `tools::minMax` calls, including calls made in various unit tests. The one exception is `grid->evalMinMax(...)`.
* `evalMinMax` has been deprecated for both tree and grid.

File: openvdb/openvdb/tools/Count.h [hunk 2]
```diff
@@ -62,6 +63,10 @@ Index64 countActiveTiles(const TreeT& tree, bool threaded = true);
 template <typename TreeT>
 Index64 memUsage(const TreeT& tree, bool threaded = true);
 
+/// @brief Return the minimum and maximum active values in this tree.
+template <typename TreeT>
+math::MinMax<typename TreeT::ValueType> minMax(const TreeT& tree, bool threaded = true);
```
Comment (Idclip) [inline openvdb/openvdb/tools/Count.h:68]: minor: a comment that this returns zeroVal<T> (i think?) for empty trees would be great

**Review by Idclip — APPROVED** [review]
This looks great, thanks a lot for your work and patience @ghurstunither.

Whilst it shouldn't be solved in this PR, the circular dependency thing is annoying - we should definitely try to fix this soon.

**Review by danrbailey — APPROVED** [review]
This is great, thank you for seeing this through!
